client-tez/src/main/java/org/apache/tez/common/RssTezUtils.java

```diff
@@ -0,0 +1,467 @@
+    } else {
+      return (int) (((1 - slowStart) * estimateMapNum + estimateReduceNum) * dynamicFactor);
+    }
+  }
+
+  public static int getRequiredShuffleServerNumber(Configuration jobConf, int mapNum, int reduceNum) {
+    int requiredShuffleServerNumber = jobConf.getInt(
+        RssTezConfig.RSS_CLIENT_ASSIGNMENT_SHUFFLE_SERVER_NUMBER,
+        RssTezConfig.RSS_CLIENT_ASSIGNMENT_SHUFFLE_SERVER_NUMBER_DEFAULT_VALUE
+    );
+    boolean enabledEstimateServer = jobConf.getBoolean(
+        RssTezConfig.RSS_ESTIMATE_SERVER_ASSIGNMENT_ENABLED,
+        RssTezConfig.RSS_ESTIMATE_SERVER_ASSIGNMENT_ENABLED_DEFAULT_VALUE
+    );
+    if (!enabledEstimateServer || requiredShuffleServerNumber > 0) {
+      return requiredShuffleServerNumber;
+    }
+    int taskConcurrency = estimateTaskConcurrency(jobConf, mapNum, reduceNum);
+    int taskConcurrencyPerServer = jobConf.getInt(RssTezConfig.RSS_ESTIMATE_TASK_CONCURRENCY_PER_SERVER,
+        RssTezConfig.RSS_ESTIMATE_TASK_CONCURRENCY_PER_SERVER_DEFAULT_VALUE);
+    return (int) Math.ceil(taskConcurrency * 1.0 / taskConcurrencyPerServer);
+  }
+
+  public static void validateRssClientConf(Configuration rssJobConf, Configuration mrJobConf) {
+    int retryMax = getInt(rssJobConf, mrJobConf, RssTezConfig.RSS_CLIENT_RETRY_MAX,
+        RssTezConfig.RSS_CLIENT_RETRY_MAX_DEFAULT_VALUE);
+    long retryIntervalMax = getLong(rssJobConf, mrJobConf, RssTezConfig.RSS_CLIENT_RETRY_INTERVAL_MAX,
+        RssTezConfig.RSS_CLIENT_RETRY_INTERVAL_MAX_DEFAULT_VALUE);
+    long sendCheckTimeout = getLong(rssJobConf, mrJobConf, RssTezConfig.RSS_CLIENT_SEND_CHECK_TIMEOUT_MS,
+        RssTezConfig.RSS_CLIENT_SEND_CHECK_TIMEOUT_MS_DEFAULT_VALUE);
+    if (retryIntervalMax * retryMax > sendCheckTimeout) {
+      throw new IllegalArgumentException(String.format("%s(%s) * %s(%s) should not bigger than %s(%s)",
+          RssTezConfig.RSS_CLIENT_RETRY_MAX,
+          retryMax,
+          RssTezConfig.RSS_CLIENT_RETRY_INTERVAL_MAX,
+          retryIntervalMax,
+          RssTezConfig.RSS_CLIENT_SEND_CHECK_TIMEOUT_MS,
+          sendCheckTimeout));
+    }
+  }
+
+  public static int computeShuffleId(int tezDagID, String upVertexName, String downVertexName) {
+    int upVertexId = mapVertexId(upVertexName);
+    int downVertexId = mapVertexId(downVertexName);
+    int shuffleId = computeShuffleId(tezDagID, upVertexId, downVertexId);
+    LOG.info("Compute Shuffle Id, upVertexName:{}, id:{}, downVertexName:{}, id:{}, shuffleId:{}",
+        upVertexName, upVertexId, downVertexName, downVertexId, shuffleId);
+    return shuffleId;
+  }
+
+  public static int computeShuffleId(int tezDagID, int upTezVertexID, int downTezVertexID) {
+    return tezDagID * (SHUFFLE_ID_MAGIC * SHUFFLE_ID_MAGIC)  + upTezVertexID * SHUFFLE_ID_MAGIC + downTezVertexID;
+  }
+
+  public static int computeShuffleId(int tezDagID, int tezVertexID) {
+    return tezDagID * 10000 + tezVertexID;
+  }
+
+  private static final int VERTEX_ID_MAPPING_MAX_ID = 500;
```

Comment:
Is the limit enough for Tez?

Reply:
Yes, one map-reduce is one vertex, mostly HQL does not have map-reduce task over 500, so it is enough.